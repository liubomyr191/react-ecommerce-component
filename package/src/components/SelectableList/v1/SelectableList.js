import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { withComponents } from "@reactioncommerce/components-context";
import { CustomPropTypes } from "../../../utils";

const StyledList = styled.div`
  fieldset {
    border-top: 1px solid #e6e6e6;
    border-left: none;
    border-right: none;
    border-bottom: none;
    border-top-right-radius: 2px;
    border-top-left-radius: 2px;
    padding: 0;
    margin: 0;
    .wrapper {
      border-bottom: 1px solid #e6e6e6;
      border-left: 1px solid #e6e6e6;
      border-right: 1px solid #e6e6e6;
      padding-left: 10px;
      padding-right: 10px;
      .leftAligned {
        justify-content: flex-start;
        label {
          font-weight: 600
        }
        .detail {
          font-size: 14px;
          font-family: "Source Sans Pro";
          margin-left: 2px;
        }
      }
      .amex,
      .visa {
        label {
          position: relative;
        }
        span {
          margin-right: 60px;
        }
        label span:after {
          left: 30px;
          position: absolute;
          content: " ";
          width: 38px;
          height: 24px;
          display: inline-block;
          border-radius: 2px;
          border: 1px solid #e6e6e6;
        }
      }
    }
  }
  .listAction {
    border-bottom: 1px solid #e6e6e6;
    border-left: 1px solid #e6e6e6;
    border-right: 1px solid #e6e6e6;
    padding-left: 10px;
    padding-right: 10px;
    height: 50px;
    display: flex;
    align-items: center;
  }
  > *:last-child {
    border-bottom-right-radius: 2px;
    border-bottom-left-radius: 2px;
    .wrapper:last-child {
      border-bottom-right-radius: 2px;
      border-bottom-left-radius: 2px;
    }
  }
`;

class SelectableList extends Component {
  static propTypes = {
    /**
     * If you've set up a components context using @reactioncommerce/components-context
     * (recommended), then this prop will come from there automatically. If you have not
     * set up a components context or you want to override one of the components in a
     * single spot, you can pass in the components prop directly.
     */
    components: PropTypes.shape({
      /**
       * Pass either the Reaction `SelectableItem` component or your own component
       * that takes `items` props and uses them to render a single item.
       */
      SelectableItem: CustomPropTypes.component.isRequired
    }).isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
      /**
       * The item ID
       */
      _id: PropTypes.string.isRequired
    })).isRequired,
    listAction: PropTypes.node,
    name: PropTypes.string.isRequired
  };

  render() {
    const {
      items,
      name,
      listAction,
      components: {
        SelectableItem,
        ...components
      },
      ...props
    } = this.props;

    return (
      <StyledList>
        <legend>Legend</legend>
        <fieldset>
          {items.map((item) =>
            <div className="wrapper">
              <SelectableItem
                name={name}
                key={item._id}
                item={item}
                component={components}
                {...props}
              />
            </div>)}
        </fieldset>
        {listAction ? <div className="listAction">{listAction}</div> : null}
      </StyledList>
    );
  }
}

export default withComponents(SelectableList);
