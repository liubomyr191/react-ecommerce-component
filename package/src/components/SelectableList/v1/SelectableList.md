### Overview

A `SelectableList` takes an array of `SelectableItems`.

#### Usage

A `SelectableList` can be enabled or disabled.

##### Default: Enabled

```jsx
const options = [{
  id: "111",
  label: "Standard (5-9 days)",
  value: "Standard"
},
{
  id: "222",
  label: "Priority (3-5 days)",
  value: "Priority"
}];

<SelectableList options={options} name="DefaultForm" value="Standard"/>
```

##### Disabled

Pass the `isReadOnly` prop to disable all options, while also displaying the chosen item.

```jsx
const options = [{
  id: "3321",
  label: "Standard (5-9 days)",
  value: "standard"
},
{
  id: "2332",
  label: "Priority (3-5 days)",
  value: "priority"
}];

<SelectableList isReadOnly options={options} name="DisabledForm" value="standard"/>
```

##### Add a list action (optional)

A `SelectableList` can optionally have a `listAction` at the end of the list. Pass any node to the  `listAction` to display a button, checkbox, or any other node.

```jsx
const readMore = (
    <Button title="Default" className="myBtn" isTextOnly isShortHeight>Add a shipping method</Button>
);

const options = [{
  id: "33321",
  label: "Standard (5-9 days)",
  value: "standard"
},
{
  id: "23342",
  label: "Priority (3-5 days)",
  value: "priority"
}];

<SelectableList options={options} listAction={readMore} name="listActionForm"/>
```

#### Styles

You can alter the appearance of a `SelectableList` using `isBordered` and `isLeftAligned` props, which can also be used together.

##### Bordered

Pass the `isBordered` prop to get a bordered list:

```jsx
const options = [{
  id: "331",
  label: "Standard (5-9 days)",
  detail: "Free",
  value: "standard"
},
{
  id: "232",
  label: "Priority (3-5 days)",
  value: "priority",
  detail: "$5.99"
}];

<SelectableList isBordered options={options} name="BorderedForm" value="standard"/>
```

##### Left-aligned

Pass the `isLeftAligned` prop to align the `detail` text to the left:

```jsx
const options = [{
  id: "3531",
  label: "Standard (5-9 days)",
  detail: "Free",
  value: "standard"
},
{
  id: "2352",
  label: "Priority (3-5 days)",
  value: "priority",
  detail: "$5.99"
}];

<SelectableList isLeftAligned options={options} value="standard" name="LeftAlignedBorderedForm"/>
```

#### Examples

##### Shipping address list

```jsx
const options = [{
  id: "1",
  label: "Standard (5-9 days)",
  value: "standard",
  detail: "Free"
},
{
  id: "2",
  label: "Priority (3-5 days)",
  value: "priority",
  detail: "$5.99"
},
{
  id: "3",
  label: "Express 2-day",
  value: "express",
  detail: "$12.99"
},
{
  id: "4",
  label: "Overnight expedited",
  value: "overnight",
  detail: "$24.99"
}];

<SelectableList isBordered options={options} name="ShippingForm"/>
```

##### Payment options list

```jsx

const iconVisa = require("../../../../../package/src/svg/iconVisa").default;

const options = [{
  id: "5",
  label: "American Express",
  detail: "ending in 4040",
  value: "amex",
  icon: iconVisa
},
{
  id: "6",
  label: "Cash",
  value: "cash"
}];

<SelectableList isBordered isLeftAligned options={options} name="PaymentForm"/>
```

##### Address list

```jsx
const link = (
    <Button title="Default" className="myBtn" isTextOnly isShortHeight>Edit</Button>
);

const addLink = (
    <Button title="Default" className="myBtn" isTextOnly isShortHeight>Add a new address</Button>
);

const options = [{
  id: "7",
  label: "Susan Doe, 2300 Buckwheat Ave, Salt Lake City, UT 84111 USA",
  value: "address1",
  detail: link
},
{
  id: "8",
  label: "Susan Doe, PO Box 1123, Salt Lake City, UT 84111 US",
  value: "address2",
  detail: link
},
{
  id: "9",
  label: "Johnny Doe, 2300 Buckwheat Ave, Salt Lake City, UT 84111 US",
  value: "address3",
  detail: link
}];

<SelectableList isBordered options={options} name="AddressForm" listAction={addLink} value="address2" />
```