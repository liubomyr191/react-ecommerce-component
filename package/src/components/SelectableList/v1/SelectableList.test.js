import React from "react";
import renderer from "react-test-renderer";
import mockComponents from "../../../tests/mockComponents";
import SelectableList from "./SelectableList";

test("basic snapshot", () => {
  const items = [{
    _id: "1",
    label: "Priority shipping",
    detail: "$12.00"

  },
  {
    _id: "2",
    label: "Expedited shipping",
    detail: "$5.00"
  },
  {
    _id: "3",
    label: "Free shipping",
    detail: "$0.00"
  }];
  const component = renderer.create(<SelectableList items={items} components={mockComponents} name="Form" />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
