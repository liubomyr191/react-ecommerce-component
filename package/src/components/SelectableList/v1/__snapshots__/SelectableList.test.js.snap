// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
.c0 {
  width: 100%;
}

.c0 fieldset {
  border-top: 1px solid #e6e6e6;
  border-left: none;
  border-right: none;
  border-bottom: none;
  border-top-right-radius: 2px;
  border-top-left-radius: 2px;
  padding: 0;
  margin: 0;
}

.c0 fieldset .wrapper {
  border-bottom: 1px solid #e6e6e6;
  border-left: 1px solid #e6e6e6;
  border-right: 1px solid #e6e6e6;
  padding: 0 10px;
}

.c0 fieldset .wrapper .leftAligned {
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c0 fieldset .wrapper .leftAligned label {
  font-weight: 700;
}

.c0 fieldset .wrapper .leftAligned .detail {
  font-size: 14px;
  font-family: 'Source Sans Pro','Helvetica Neue',Helvetica,sans-serif;
  margin-left: 2px;
}

.c0 fieldset .wrapper .amex label,
.c0 fieldset .wrapper .visa label {
  position: relative;
}

.c0 fieldset .wrapper .amex span,
.c0 fieldset .wrapper .visa span {
  margin: 0 10px 0 0;
}

.c0 fieldset .wrapper .amex label span:after,
.c0 fieldset .wrapper .visa label span:after {
  left: 30px;
  position: absolute;
  content: " ";
  width: 38px;
  height: 24px;
  display: inline-block;
  border-radius: 2px;
  border: 1px solid #e6e6e6;
}

.c0 .listAction {
  border-bottom: 1px solid #e6e6e6;
  border-left: 1px solid #e6e6e6;
  border-right: 1px solid #e6e6e6;
  padding: 0 10px;
  height: 50px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
}

.c0 > *:last-child {
  border-bottom-right-radius: 2px;
  border-bottom-left-radius: 2px;
}

.c0 > *:last-child .wrapper:last-child {
  border-bottom-right-radius: 2px;
  border-bottom-left-radius: 2px;
}

@media (max-width:768px) {
  .c0 .listAction {
    height: 60px;
  }
}

<div
  className="c0"
>
  <fieldset>
    <div
      className="wrapper"
    >
      SelectableItem(undefined)
    </div>
    <div
      className="wrapper"
    >
      SelectableItem(undefined)
    </div>
    <div
      className="wrapper"
    >
      SelectableItem(undefined)
    </div>
  </fieldset>
</div>
`;
